File: Practice/Arrays/CommonElements.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class CommonElements {
    public int[] solve(int[] A, int[] B) {

        HashMap<Integer, Integer> hm = new HashMap<Integer, Integer>();
        List<Integer> l = new ArrayList<Integer>();

        for (int i = 0; i < A.length; i++) {
            int val = hm.getOrDefault(A[i], 0);
            hm.put(A[i], val + 1);
        }

        for (int j = 0; j < B.length; j++) {
            if (hm.size() == 0)
                break;
            if (hm.containsKey(B[j])) {
                if (hm.get(B[j]) == 0) {
                    hm.remove(B[j]);
                } else {
                    l.add(B[j]);
                    hm.put(B[j], hm.get(B[j]) - 1);
                }
            }
        }

        int[] output = new int[l.size()];
        for (int y = 0; y < l.size(); y++) {
            output[y] = l.get(y);
        }
        return output;
    }
}

File: Practice/Hashing/CommonElements.js
function CommonElements(A, B) {
  const hm = new Map();
  let output = [];
  for (let i = 0; i < B.length; i++) {
    hm.set(B[i], hm.has(B[i]) ? hm.get(B[i]) + 1 : 1);
  }

  for (let i = 0; i < A.length; i++) {
    if (hm.has(A[i])) {
      output.push(A[i]);
      hm.set(A[i], hm.get(A[i]) - 1);
      if (hm.get(A[i]) == 0) hm.delete(A[i]);
    }
  }

  return output;
}

A = [1, 2, 2, 1];
B = [2, 3, 1, 2];
console.log(CommonElements(A, B));
